# **Employee Turnover Costs Are More Expensive Than We Think**

Presented By: Alireza & Krunal

---

**Introduction:**

One of the most critical issues facing organizations today is how to retain the employees they want to keep. Anually, around one quarter of all U.S. workers quit their jobs, and in some industries the turnover rate is considerably higher. Nowadays, the company that has more engaged and productive employees will have a better chance of winning market competition. For this reason, companies can not lose important employees and when that begins to happen we need to understand why, to prevent this from happening.
A large company named AEKK, with around 4000 employees, face with high rate of human resources turnover. Anually, around 15% of its employees leave the company and need to be replaced with the talent pool available in the job market. management believes that this level of attrition (employees leaving, either on their own or because they got fired) is bad for the company. because of the following reasons:
•	The former employees’ projects get delayed, which makes it difficult to meet timelines, resulting in a reputation loss among consumers and partners
•	A sizeable department has to be maintained, for the purposes of recruiting new talent
•	More often than not, the new employees have to be trained for the job and/or given time to acclimatise themselves to the company




**Our Dataset:**

The dataset contains 28 features which are to be used for predicting the probability of whether an employee will leave the company or not. The features were part of three data files - one was collected by an employee survey, which measured job satisfaction and related factors. The second was a manager survey, which measured job involvement and performance rating of the employees. The final file was a General Data file with various details of the employees like Age, Travel Frequency, Department, etc.



There are 3 data files that we will be using for this case study:



**1.general_data.csv**

Age: Age of the employee

Attrition: Whether the employee has left the organisation

BusinessTravel: How frequent the employee travelled for 
business in the last year

Department: Employee's department

DistanceFromHome: Distance between Office and Employee's home (in km)

Education: Employee's level of education (1: 'Below College' , 2: 'College' , 3: 'Bachelor's Degree' , 4 : 'Masters Degree' , 5. 'Doctorate')

EducationField: Employee's field of education

EmployeeCount: Employee count

EmployeeID: Unique Employee ID

Gender: Employee's gender

JobLevel: Employee's job level on a scale of 1 to 5

JobRole: Employee's role title

MaritalStatus: Employee's marital status

MonthlyIncome: Employee's monthly income (in Rupees per month)

NumCompaniesWorked: Total number of companies the employee has worked for

Over18: Whether the employee is above 18 years of age

PercentSalaryHike: Employee's salary hike last year (in percentage points)

StandardHours: Employee's standard working hours (duration)

StockOptionLevel: Employee's stock option level

TotalWorkingYears: Employee's total number of working years (entire life)

TrainingTimesLastYear: Number of times employee attended training last year

YearsAtCompany: Employee's total number of working years (in the company)

YearsSinceLastPromotion: Employee's number of years since last promotion

YearsWithCurrManager: Employee's number of years working under current manager

**2.employee_survey_data.csv**

EmployeeID: Unique Employee ID

EnvironmentSatisfaction: Employee's Work Environment Satisfaction Level (1:'Low', 2:'Medium', 3:'High', 4: 'Very High')

JobSatisfaction: Employee's Job Satisfaction Level (1:'Low', 2:'Medium', 3:'High', 4 :'Very High')

WorkLifeBalance: Employee's Work Life Balance Rating Level (1:'Low' , 2:'Medium' , 3:'High' , 4 :'Very High')

**3.manager_survey_data.csv**

EmployeeID: Unique Employee ID

JobInvolvement: Employee's Job Involvement Level (1:'Low',2:'Medium', 3:'High', 4:'Very High')

PerformanceRating: Employee's performance rating last year

### 1.Data Exploring

In [100]:
# Mounting Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [101]:
pip install dmba

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [102]:
# Import libraries
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from matplotlib import pyplot as plt
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import classification_report
from dmba import plotDecisionTree, classificationSummary, regressionSummary

from sklearn import preprocessing
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

from itertools import chain
import torch
import numpy as np
import torch.nn as nn
import torch.optim as optim

In [103]:
# Read general data and show the head rows of that
General = pd.read_csv("/content/drive/MyDrive/DL_data/general_data.csv")
General.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [104]:
General.describe().T

count          mean           std      min  \
Age                      4410.0     36.923810      9.133301     18.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0   
Education                4410.0      2.912925      1.023933      1.0   
EmployeeCount            4410.0      1.000000      0.000000      1.0   
EmployeeID               4410.0   2205.500000   1273.201673      1.0   
JobLevel                 4410.0      2.063946      1.106689      1.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0   
NumCompaniesWorked       4391.0      2.694830      2.498887      0.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0   
StandardHours            4410.0      8.000000      0.000000      8.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0   
TotalWorkingYears        4401.0     11.279936      7.782222      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0   

                              25%      50%       75%       max  
Age                         30.00     36.0     43.00      60.0  
DistanceFromHome             2.00      7.0     14.00      29.0  
Education                    2.00      3.0      4.00       5.0  
EmployeeCount                1.00      1.0      1.00       1.0  
EmployeeID                1103.25   2205.5   3307.75    4410.0  
JobLevel                     1.00      2.0      3.00       5.0  
MonthlyIncome            29110.00  49190.0  83800.00  199990.0  
NumCompaniesWorked           1.00      2.0      4.00       9.0  
PercentSalaryHike           12.00     14.0     18.00      25.0  
StandardHours                8.00      8.0      8.00       8.0  
StockOptionLevel             0.00      1.0      1.00       3.0  
TotalWorkingYears            6.00     10.0     15.00      40.0  
TrainingTimesLastYear        2.00      3.0      3.00       6.0  
YearsAtCompany               3.00      5.0      9.00      40.0  
YearsSinceLastPromotion      0.00      1.0      3.00      15.0  
YearsWithCurrManager         2.00      3.0      7.00      17.0

In [105]:
# Read emplyee data and show the head rows of that
Employee = pd.read_csv("/content/drive/MyDrive/DL_data/employee_survey_data.csv")
Employee.head()

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance
0           1                      3.0              4.0              2.0
1           2                      3.0              2.0              4.0
2           3                      2.0              2.0              1.0
3           4                      4.0              4.0              3.0
4           5                      4.0              1.0              3.0

In [106]:
Employee.describe()

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance
count  4410.000000              4385.000000      4390.000000      4372.000000
mean   2205.500000                 2.723603         2.728246         2.761436
std    1273.201673                 1.092756         1.101253         0.706245
min       1.000000                 1.000000         1.000000         1.000000
25%    1103.250000                 2.000000         2.000000         2.000000
50%    2205.500000                 3.000000         3.000000         3.000000
75%    3307.750000                 4.000000         4.000000         3.000000
max    4410.000000                 4.000000         4.000000         4.000000

In [107]:
# Read manager data and show the head rows of that
Manager= pd.read_csv("/content/drive/MyDrive/DL_data/manager_survey_data.csv")
Manager.head()

EmployeeID  JobInvolvement  PerformanceRating
0           1               3                  3
1           2               2                  4
2           3               3                  3
3           4               2                  3
4           5               3                  3

In [108]:
Manager.describe()

EmployeeID  JobInvolvement  PerformanceRating
count  4410.000000     4410.000000        4410.000000
mean   2205.500000        2.729932           3.153741
std    1273.201673        0.711400           0.360742
min       1.000000        1.000000           3.000000
25%    1103.250000        2.000000           3.000000
50%    2205.500000        3.000000           3.000000
75%    3307.750000        3.000000           3.000000
max    4410.000000        4.000000           4.000000

In [109]:
#Combining the three files into one dataframe based on EmployeeID
df = pd.merge(pd.merge(General, Employee, on = 'EmployeeID'), Manager, on = 'EmployeeID')
df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   TotalWorkingYears TrainingTimesLastYear YearsAtCompany  \
0                1.0                     6              1   
1                6.0                     3              5   
2                5.0                     2              5   
3               13.0                     5              8   
4                9.0                     2              6   

   YearsSinceLastPromotion  YearsWithCurrManager EnvironmentSatisfaction  \
0                        0                     0                     3.0   
1                        1                     4                     3.0   
2                        0                     3                     2.0   
3                        7                     5                     4.0   
4                        0                     4                     4.0   

   JobSatisfaction  WorkLifeBalance  JobInvolvement  PerformanceRating  
0              4.0              2.0               3                  3  
1              2.0              4.0               2                  4  
2              2.0              1.0               3                  3  
3              4.0              3.0               2                  3  
4              1.0              3.0               3                  3  

[5 rows x 29 columns]

In [110]:
df.describe().T

count          mean           std      min  \
Age                      4410.0     36.923810      9.133301     18.0   
DistanceFromHome         4410.0      9.192517      8.105026      1.0   
Education                4410.0      2.912925      1.023933      1.0   
EmployeeCount            4410.0      1.000000      0.000000      1.0   
EmployeeID               4410.0   2205.500000   1273.201673      1.0   
JobLevel                 4410.0      2.063946      1.106689      1.0   
MonthlyIncome            4410.0  65029.312925  47068.888559  10090.0   
NumCompaniesWorked       4391.0      2.694830      2.498887      0.0   
PercentSalaryHike        4410.0     15.209524      3.659108     11.0   
StandardHours            4410.0      8.000000      0.000000      8.0   
StockOptionLevel         4410.0      0.793878      0.851883      0.0   
TotalWorkingYears        4401.0     11.279936      7.782222      0.0   
TrainingTimesLastYear    4410.0      2.799320      1.288978      0.0   
YearsAtCompany           4410.0      7.008163      6.125135      0.0   
YearsSinceLastPromotion  4410.0      2.187755      3.221699      0.0   
YearsWithCurrManager     4410.0      4.123129      3.567327      0.0   
EnvironmentSatisfaction  4385.0      2.723603      1.092756      1.0   
JobSatisfaction          4390.0      2.728246      1.101253      1.0   
WorkLifeBalance          4372.0      2.761436      0.706245      1.0   
JobInvolvement           4410.0      2.729932      0.711400      1.0   
PerformanceRating        4410.0      3.153741      0.360742      3.0   

                              25%      50%       75%       max  
Age                         30.00     36.0     43.00      60.0  
DistanceFromHome             2.00      7.0     14.00      29.0  
Education                    2.00      3.0      4.00       5.0  
EmployeeCount                1.00      1.0      1.00       1.0  
EmployeeID                1103.25   2205.5   3307.75    4410.0  
JobLevel                     1.00      2.0      3.00       5.0  
MonthlyIncome            29110.00  49190.0  83800.00  199990.0  
NumCompaniesWorked           1.00      2.0      4.00       9.0  
PercentSalaryHike           12.00     14.0     18.00      25.0  
StandardHours                8.00      8.0      8.00       8.0  
StockOptionLevel             0.00      1.0      1.00       3.0  
TotalWorkingYears            6.00     10.0     15.00      40.0  
TrainingTimesLastYear        2.00      3.0      3.00       6.0  
YearsAtCompany               3.00      5.0      9.00      40.0  
YearsSinceLastPromotion      0.00      1.0      3.00      15.0  
YearsWithCurrManager         2.00      3.0      7.00      17.0  
EnvironmentSatisfaction      2.00      3.0      4.00       4.0  
JobSatisfaction              2.00      3.0      4.00       4.0  
WorkLifeBalance              2.00      3.0      3.00       4.0  
JobInvolvement               2.00      3.0      3.00       4.0  
PerformanceRating            3.00      3.0      3.00       4.0

In [111]:
df.shape

(4410, 29)

In [112]:
# Examine variable type
df.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeID                   int64
Gender                      object
JobLevel                     int64
JobRole                     object
MaritalStatus               object
MonthlyIncome                int64
NumCompaniesWorked         float64
Over18                      object
PercentSalaryHike            int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears          float64
TrainingTimesLastYear        int64
YearsAtCompany               int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
EnvironmentSatisfaction    float64
JobSatisfaction            float64
WorkLifeBalance            float64
JobInvolvement               int64
PerformanceRating   

In [113]:
# Histogram of insurance expenses
sns.countplot(x='Attrition',data=df)

In [114]:
plt.figure(figsize=(18, 10))
corr = df.corr()
sns.heatmap(corr,cmap ="YlGnBu",cbar=True,annot=True,fmt=".1f")
plt.show()

In [115]:
corrdf = corr.where(np.triu(np.ones(corr.shape), k=1).astype(np.bool))
corrdf = corrdf.unstack().reset_index()
corrdf.columns = ['feature1', 'feature2', 'Correlation']
corrdf.dropna(subset = ['Correlation'], inplace = True)
corrdf['Correlation'] = round(corrdf['Correlation'], 2)
corrdf['Correlation'] = abs(corrdf['Correlation'])
matrix= corrdf.sort_values(by = 'Correlation', ascending = False)
high_corr = matrix[matrix['Correlation']>0.5]
high_corr

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


feature1                 feature2  Correlation
428        PerformanceRating        PercentSalaryHike         0.77
328     YearsWithCurrManager           YearsAtCompany         0.77
231        TotalWorkingYears                      Age         0.68
284           YearsAtCompany        TotalWorkingYears         0.63
307  YearsSinceLastPromotion           YearsAtCompany         0.62
329     YearsWithCurrManager  YearsSinceLastPromotion         0.51

**Conclusion from Correlation Matrix :**

**Percentage Salary hike strongly correlates with performance rating:** The people who have good rating get paid better.

**YearsAtCompany correlates with YearsWithCurrManager:** This means the employees never get reassigned

**Age correlates with the number of year worked in:** The older you are the more you have been working.

**TotalNumberWorkingYears correlates with YearsAtCompany:** More working years of employees is in the company. 

**YearsAtCompany correlates with YearsSinceLastPromotion:** The employee are not promoted based on experience and the number of years they have contributed 

**YearsSinceLastPromotion correlates with YearsWithCurrManager:** This means the employees never get promoted maybe the manager is the reason and they don't let people get promoted.



Now further we will be analyzing the categorical variable through count plot, For this we have created a list feature_cat having the required categorical variable, Lets see how the plot comes out and what can be inferred from this.

In [116]:
sns.countplot(x ='BusinessTravel',data=df,hue="Attrition")
print('\nAttrition rate by BusinessTravel \n')
print(round(df['BusinessTravel'][df['Attrition'] == 'Yes'].value_counts()/df['BusinessTravel'].value_counts()*100,2))


Attrition rate by BusinessTravel 

Travel_Rarely        14.96
Travel_Frequently    24.91
Non-Travel            8.00
Name: BusinessTravel, dtype: float64


In [117]:
sns.countplot(x ='Department',data=df,hue="Attrition")
print('\nAttrition rate by Department \n')
print(round(df['Department'][df['Attrition'] == 'Yes'].value_counts()/df['Department'].value_counts()*100,2))


Attrition rate by Department 

Research & Development    15.71
Sales                     15.02
Human Resources           30.16
Name: Department, dtype: float64


In [118]:
sns.countplot(x ='EducationField',data=df,hue="Attrition")
plt.xticks(rotation=90)
print('\nAttrition rate by EducationField \n')
print(round(df['EducationField'][df['Attrition'] == 'Yes'].value_counts()/df['EducationField'].value_counts()*100,2))


Attrition rate by EducationField 

Human Resources     40.74
Life Sciences       16.67
Marketing           15.72
Medical             16.16
Other               12.20
Technical Degree    11.36
Name: EducationField, dtype: float64


In [119]:
sns.countplot(x ='Gender',data=df,hue="Attrition")
print('\nAttrition rate by Gender \n')
print(round(df['Gender'][df['Attrition'] == 'Yes'].value_counts()/df['Gender'].value_counts()*100,2))


Attrition rate by Gender 

Male      16.67
Female    15.31
Name: Gender, dtype: float64


In [120]:
sns.countplot(x ='JobRole',data=df,hue="Attrition")
plt.xticks(rotation=90)
print('\nAttrition rate by JobRole \n')
print(round(df['JobRole'][df['Attrition'] == 'Yes'].value_counts()/df['JobRole'].value_counts()*100,2))


Attrition rate by JobRole 

Healthcare Representative    14.50
Human Resources              13.46
Laboratory Technician        16.22
Manager                      13.73
Manufacturing Director       11.03
Research Director            23.75
Research Scientist           18.15
Sales Executive              16.87
Sales Representative         14.46
Name: JobRole, dtype: float64


In [121]:
sns.countplot(x ='MaritalStatus',data=df,hue="Attrition")
print('\nAttrition rate by MaritalStatus \n')
print(round(df['MaritalStatus'][df['Attrition'] == 'Yes'].value_counts()/df['MaritalStatus'].value_counts()*100,2))


Attrition rate by MaritalStatus 

Divorced    10.09
Married     12.48
Single      25.53
Name: MaritalStatus, dtype: float64


From this plot we can see that for different categories what are the number of people who stayed and who left. Observetions are as follows:

Employess those travel very frequently are most likely to leave. From the above graph it seems like company need to check its policy on travelling.

More than 30% from HR department also leaves the company.

Single Employess has a high tendency to leave.

Employees who work a research director have a high tendency to leave.

### 2.Data Preprocessing

In [122]:
# Change categorical variables to "category"
df['Attrition'] = df['Attrition'].astype('category')
df['BusinessTravel'] = df['BusinessTravel'].astype('category')
df['Department'] = df['Department'].astype('category')
df['Gender'] = df['Gender'].astype('category')
df['JobRole'] = df['JobRole'].astype('category')
df['MaritalStatus'] = df['MaritalStatus'].astype('category')
df['Over18'] = df['Over18'].astype('category')
df['EducationField'] = df['EducationField'].astype('category')

In [123]:
# Examine variable type
df.dtypes

Age                           int64
Attrition                  category
BusinessTravel             category
Department                 category
DistanceFromHome              int64
Education                     int64
EducationField             category
EmployeeCount                 int64
EmployeeID                    int64
Gender                     category
JobLevel                      int64
JobRole                    category
MaritalStatus              category
MonthlyIncome                 int64
NumCompaniesWorked          float64
Over18                     category
PercentSalaryHike             int64
StandardHours                 int64
StockOptionLevel              int64
TotalWorkingYears           float64
TrainingTimesLastYear         int64
YearsAtCompany                int64
YearsSinceLastPromotion       int64
YearsWithCurrManager          int64
EnvironmentSatisfaction     float64
JobSatisfaction             float64
WorkLifeBalance             float64
JobInvolvement              

In [124]:
#determine columns with blank/missing data
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeID                  0
Gender                      0
JobLevel                    0
JobRole                     0
MaritalStatus               0
MonthlyIncome               0
NumCompaniesWorked         19
Over18                      0
PercentSalaryHike           0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           9
TrainingTimesLastYear       0
YearsAtCompany              0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
EnvironmentSatisfaction    25
JobSatisfaction            20
WorkLifeBalance            38
JobInvolvement              0
PerformanceRating           0
dtype: int64

In [125]:
#Replace null fields with median because they are all integer and check again
df.fillna(df.median(), inplace=True)
df.isnull().sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  


Age                        0
Attrition                  0
BusinessTravel             0
Department                 0
DistanceFromHome           0
Education                  0
EducationField             0
EmployeeCount              0
EmployeeID                 0
Gender                     0
JobLevel                   0
JobRole                    0
MaritalStatus              0
MonthlyIncome              0
NumCompaniesWorked         0
Over18                     0
PercentSalaryHike          0
StandardHours              0
StockOptionLevel           0
TotalWorkingYears          0
TrainingTimesLastYear      0
YearsAtCompany             0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
EnvironmentSatisfaction    0
JobSatisfaction            0
WorkLifeBalance            0
JobInvolvement             0
PerformanceRating          0
dtype: int64

In [126]:
# checking any duplicated employee entries
CheckDup = df['EmployeeID'].nunique() - df['EmployeeID'].count()
print('No. of duplicated Employee records :')
print(CheckDup)

No. of duplicated Employee records :
0


In [127]:
#Check columns with just 1 unique value
uniquecount = df.nunique()
print(uniquecount)

Age                          43
Attrition                     2
BusinessTravel                3
Department                    3
DistanceFromHome             29
Education                     5
EducationField                6
EmployeeCount                 1
EmployeeID                 4410
Gender                        2
JobLevel                      5
JobRole                       9
MaritalStatus                 3
MonthlyIncome              1349
NumCompaniesWorked           10
Over18                        1
PercentSalaryHike            15
StandardHours                 1
StockOptionLevel              4
TotalWorkingYears            40
TrainingTimesLastYear         7
YearsAtCompany               37
YearsSinceLastPromotion      16
YearsWithCurrManager         18
EnvironmentSatisfaction       4
JobSatisfaction               4
WorkLifeBalance               4
JobInvolvement                4
PerformanceRating             2
dtype: int64


In [128]:
#Remove columns identified above and Employee ID which is not relevant
df = df.drop(columns = ['EmployeeCount', 'Over18','StandardHours','EmployeeID','JobInvolvement','JobSatisfaction','WorkLifeBalance','EnvironmentSatisfaction','TrainingTimesLastYear','StockOptionLevel','MonthlyIncome','DistanceFromHome'])

In [129]:
df.head()

Age Attrition     BusinessTravel              Department  Education  \
0   51        No      Travel_Rarely                   Sales          2   
1   31       Yes  Travel_Frequently  Research & Development          1   
2   32        No  Travel_Frequently  Research & Development          4   
3   38        No         Non-Travel  Research & Development          5   
4   32        No      Travel_Rarely  Research & Development          1   

  EducationField  Gender  JobLevel                    JobRole MaritalStatus  \
0  Life Sciences  Female         1  Healthcare Representative       Married   
1  Life Sciences  Female         1         Research Scientist        Single   
2          Other    Male         4            Sales Executive       Married   
3  Life Sciences    Male         3            Human Resources       Married   
4        Medical    Male         1            Sales Executive        Single   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                 1.0                 11                1.0               1   
1                 0.0                 23                6.0               5   
2                 1.0                 15                5.0               5   
3                 3.0                 11               13.0               8   
4                 4.0                 12                9.0               6   

   YearsSinceLastPromotion  YearsWithCurrManager  PerformanceRating  
0                        0                     0                  3  
1                        1                     4                  4  
2                        0                     3                  3  
3                        7                     5                  3  
4                        0                     4                  3

In [130]:
df.shape

(4410, 17)

### 3.Decision Tree Model

In [131]:
df.dtypes

Age                           int64
Attrition                  category
BusinessTravel             category
Department                 category
Education                     int64
EducationField             category
Gender                     category
JobLevel                      int64
JobRole                    category
MaritalStatus              category
NumCompaniesWorked          float64
PercentSalaryHike             int64
TotalWorkingYears           float64
YearsAtCompany                int64
YearsSinceLastPromotion       int64
YearsWithCurrManager          int64
PerformanceRating             int64
dtype: object

In [132]:
# Create dummy variables
df = pd.get_dummies(df, columns=['BusinessTravel','Department','EducationField','Gender','JobRole','MaritalStatus'],drop_first=True)
df

Age Attrition  Education  JobLevel  NumCompaniesWorked  \
0      51        No          2         1                 1.0   
1      31       Yes          1         1                 0.0   
2      32        No          4         4                 1.0   
3      38        No          5         3                 3.0   
4      32        No          1         1                 4.0   
...   ...       ...        ...       ...                 ...   
4405   42        No          4         1                 3.0   
4406   29        No          4         1                 2.0   
4407   25        No          2         2                 0.0   
4408   42        No          2         1                 0.0   
4409   40        No          3         2                 0.0   

      PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                    11                1.0               1   
1                    23                6.0               5   
2                    15                5.0               5   
3                    11               13.0               8   
4                    12                9.0               6   
...                 ...                ...             ...   
4405                 17               10.0               3   
4406                 15               10.0               3   
4407                 20                5.0               4   
4408                 14               10.0               9   
4409                 12               10.0              21   

      YearsSinceLastPromotion  YearsWithCurrManager  ...  \
0                           0                     0  ...   
1                           1                     4  ...   
2                           0                     3  ...   
3                           7                     5  ...   
4                           0                     4  ...   
...                       ...                   ...  ...   
4405                        0                     2  ...   
4406                        0                     2  ...   
4407                        1                     2  ...   
4408                        7                     8  ...   
4409                        3                     9  ...   

      JobRole_Human Resources  JobRole_Laboratory Technician  JobRole_Manager  \
0                           0                              0                0   
1                           0                              0                0   
2                           0                              0                0   
3                           1                              0                0   
4                           0                              0                0   
...                       ...                            ...              ...   
4405                        0                              0                0   
4406                        0                              1                0   
4407                        0                              0                0   
4408                        0                              1                0   
4409                        0                              1                0   

      JobRole_Manufacturing Director  JobRole_Research Director  \
0                                  0                          0   
1                                  0                          0   
2                                  0                          0   
3                                  0                          0   
4                                  0                          0   
...                              ...                        ...   
4405                               0                          0   
4406                               0                          0   
4407                               0                          0   
4408                               0                          0   
4409                               0                         

In [133]:
# Apply standardization
numeric_variables = df[['Age','Education', 'JobLevel','NumCompaniesWorked','PercentSalaryHike','TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager','PerformanceRating']]
scaler_s = StandardScaler().fit(numeric_variables)
standard_variables = scaler_s.transform(numeric_variables)
print(standard_variables)
df[['Age','Education', 'JobLevel','NumCompaniesWorked','PercentSalaryHike','TotalWorkingYears','YearsAtCompany','YearsSinceLastPromotion','YearsWithCurrManager','PerformanceRating']] = standard_variables
df

[[ 1.54136892 -0.89168825 -0.96148639 ... -0.67914568 -1.15593471
  -0.42623002]
 [-0.64866811 -1.86842575 -0.96148639 ... -0.36871529 -0.03451975
   2.34615106]
 [-0.53916626  1.06178675  1.74961015 ... -0.67914568 -0.31487349
  -0.42623002]
 ...
 [-1.30567922 -0.89168825 -0.05778755 ... -0.36871529 -0.59522723
   2.34615106]
 [ 0.55585225 -0.89168825 -0.96148639 ...  1.49386709  1.08689522
  -0.42623002]
 [ 0.33684855  0.08504925 -0.05778755 ...  0.25214551  1.36724896
  -0.42623002]]


Age Attrition  Education  JobLevel  NumCompaniesWorked  \
0     1.541369        No  -0.891688 -0.961486           -0.678464   
1    -0.648668       Yes  -1.868426 -0.961486           -1.079486   
2    -0.539166        No   1.061787  1.749610           -0.678464   
3     0.117845        No   2.038524  0.845911            0.123580   
4    -0.539166        No  -1.868426 -0.961486            0.524602   
...        ...       ...        ...       ...                 ...   
4405  0.555852        No   1.061787 -0.961486            0.123580   
4406 -0.867672        No   1.061787 -0.961486           -0.277442   
4407 -1.305679        No  -0.891688 -0.057788           -1.079486   
4408  0.555852        No  -0.891688 -0.961486           -1.079486   
4409  0.336849        No   0.085049 -0.057788           -1.079486   

      PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0             -1.150554          -1.322079       -0.981014   
1              2.129306          -0.678877       -0.327893   
2             -0.057267          -0.807517       -0.327893   
3             -1.150554           0.221606        0.161947   
4             -0.877232          -0.292956       -0.164613   
...                 ...                ...             ...   
4405           0.489376          -0.164315       -0.654454   
4406          -0.057267          -0.164315       -0.654454   
4407           1.309341          -0.807517       -0.491174   
4408          -0.330589          -0.164315        0.325228   
4409          -0.877232          -0.164315        2.284590   

      YearsSinceLastPromotion  YearsWithCurrManager  ...  \
0                   -0.679146             -1.155935  ...   
1                   -0.368715             -0.034520  ...   
2                   -0.679146             -0.314873  ...   
3                    1.493867              0.245834  ...   
4                   -0.679146             -0.034520  ...   
...                       ...                   ...  ...   
4405                -0.679146             -0.595227  ...   
4406                -0.679146             -0.595227  ...   
4407                -0.368715             -0.595227  ...   
4408                 1.493867              1.086895  ...   
4409                 0.252146              1.367249  ...   

      JobRole_Human Resources  JobRole_Laboratory Technician  JobRole_Manager  \
0                           0                              0                0   
1                           0                              0                0   
2                           0                              0                0   
3                           1                              0                0   
4                           0                              0                0   
...                       ...                            ...              ...   
4405                        0                              0                0   
4406                        0                              1                0   
4407                        0                              0                0   
4408                        0                              1                0   
4409                        0                              1                0   

      JobRole_Manufacturing Director  JobRole_Research Director  \
0                                  0                          0   
1                                  0                          0   
2                                  0                          0   
3                                  0                          0   
4                                  0                          0   
...                              ...                        ...   
4405                               0                          0   
4406                               0                          0   
4407                               0                          0   
4408                               0                          0   
4409  

In [134]:
# Partition the data
target = df['Attrition']
predictors = df.drop(['Attrition'], axis=1)
predictors_train, predictors_test, target_train, target_test = train_test_split(predictors, target, test_size = 0.3, random_state = 0)
print(predictors_train.shape, predictors_test.shape, target_train.shape, target_test.shape)

(3087, 30) (1323, 30) (3087,) (1323,)


In [135]:
DecTree = DecisionTreeClassifier(random_state=0)
DecTree.fit(predictors_train,target_train)
print("The classification report for Desicion Tree model:\n", classification_report(target_test, DecTree.predict(predictors_test)))
print("The confusion matrix for Desicion Tree model:\n",confusion_matrix(target_test, DecTree.predict(predictors_test)))

The classification report for Desicion Tree model:
               precision    recall  f1-score   support

          No       0.99      0.99      0.99      1127
         Yes       0.93      0.97      0.95       196

    accuracy                           0.98      1323
   macro avg       0.96      0.98      0.97      1323
weighted avg       0.98      0.98      0.98      1323

The confusion matrix for Desicion Tree model:
 [[1112   15]
 [   6  190]]


The model predicts 205 cases of attrition, out of which 190 employees actually have left the company. There are 15 false negative, what gives us good precision score = 0.93. Model also didn't show just 6 of employees who actually left the company, hence the best recall score = 0.97

### 4.Logistic Regression model

In [136]:
# creating model and fitting training data
logreg = LogisticRegression(solver='liblinear', random_state = 1000)
logreg.fit(predictors_train,target_train)

LogisticRegression(random_state=1000, solver='liblinear')

In [137]:
# Obtain the predictions from our logistic regression model:
y_pred = logreg.predict(predictors_test)

# Compute predicted probabilities: y_pred_prob
y_pred_prob = logreg.predict_proba(predictors_test)[:,1]

#classification report and confusion matrix
print("The classification report for logestic regression model:\n", classification_report(target_test, y_pred))
print("The confusion matrix for logestic regression model:\n",confusion_matrix(target_test, y_pred))


The classification report for logestic regression model:
               precision    recall  f1-score   support

          No       0.86      0.99      0.92      1127
         Yes       0.57      0.06      0.11       196

    accuracy                           0.85      1323
   macro avg       0.72      0.53      0.52      1323
weighted avg       0.82      0.85      0.80      1323

The confusion matrix for logestic regression model:
 [[1118    9]
 [ 184   12]]


The model predicts 21 cases of attrition, out of which 12 employees actually have left the company. There are 9 false negative, what gives us pretty poor precision score = 0.57. Model also didn't catch 184 of employees who actually left the company, hence very low recall score = 0.06

In [138]:
# creating model and fitting training data
logregSa = LogisticRegression(solver='saga', random_state = 1000)
logregSa.fit(predictors_train,target_train)

LogisticRegression(random_state=1000, solver='saga')

In [139]:
# Obtain the predictions from our logistic regression model:
y_pred = logregSa.predict(predictors_test)

# Compute predicted probabilities: y_pred_prob
y_pred_prob = logregSa.predict_proba(predictors_test)[:,1]

#classification report and confusion matrix
print("The classification report for logestic regression model:\n", classification_report(target_test, y_pred))
print("The confusion matrix for logestic regression model:\n",confusion_matrix(target_test, y_pred))

The classification report for logestic regression model:
               precision    recall  f1-score   support

          No       0.86      0.99      0.92      1127
         Yes       0.57      0.06      0.11       196

    accuracy                           0.85      1323
   macro avg       0.72      0.53      0.52      1323
weighted avg       0.82      0.85      0.80      1323

The confusion matrix for logestic regression model:
 [[1118    9]
 [ 184   12]]


### 4.Neural network prediction and evaluation

In [140]:
# Build a neural network on training data
class neural_network(nn.Module):
    def __init__(self,  in_size, hidden_size, out_size):
        super().__init__()
        self.network = nn.Sequential(
          nn.Linear(in_size, hidden_size),
          nn.ReLU(),
          nn.Linear(hidden_size, hidden_size),
          nn.ReLU(),
          nn.Linear(hidden_size, out_size))

    def forward(self, x):
        out = self.network(x)
        return out

In [141]:
# Encode label to numeric
label_encoder = preprocessing.LabelEncoder()
new_label = label_encoder.fit_transform(target_train)
new_label[:5], target_train[:5]

(array([0, 0, 0, 0, 0]), 1087    No
 1621    No
 1804    No
 744     No
 686     No
 Name: Attrition, dtype: category
 Categories (2, object): ['No', 'Yes'])

In [142]:
# Create tensors from pandas dataframe
predictors_train_tensor = torch.tensor(predictors_train.values)
target_train_tensor = torch.tensor(label_encoder.transform(target_train.values))
predictors_test_tensor = torch.tensor(predictors_test.values)
target_test_tensor = torch.tensor(label_encoder.transform(target_test.values))

# Create tensor dataset (set target variable to long int type)
train_dataset = torch.utils.data.TensorDataset(predictors_train_tensor.float(), target_train_tensor.long())
test_dataset = torch.utils.data.TensorDataset(predictors_test_tensor.float(), target_test_tensor.long())

# Define training and testing data loader, and set batch size to 64
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=64, shuffle=False)

In [143]:
# Define training loop function
def training_loop(n_epochs, optimizer, model, loss_fn, train_loader):
    for epoch in range(0, n_epochs):
        # Training Phase 
        model.train()
        loss_train = 0.0
        for inputs, labels in train_loader:

            outputs = model(inputs)
            
            loss = loss_fn(outputs, labels)
            
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            loss_train += loss.item()

        if epoch == 0 or epoch == n_epochs-1 or epoch % 10 == 0:
            print('Epoch {}, Training loss {}'.format(epoch, loss_train / len(train_loader)))

In [154]:
# Model training
torch.manual_seed(0)
model=neural_network(30,128,2)
optimizer_adam=optim.Adam(model.parameters())
loss_fn=nn.CrossEntropyLoss()
training_loop(n_epochs=50, optimizer=optimizer_adam, model=model, loss_fn=loss_fn, train_loader=train_loader)

Epoch 0, Training loss 0.48080481680072085
Epoch 10, Training loss 0.20823511663748293
Epoch 20, Training loss 0.04772534145384419
Epoch 30, Training loss 0.007742855949707481
Epoch 40, Training loss 0.002421608943093036
Epoch 49, Training loss 0.0012656550109624027


In [155]:
# Define testing function
def test(model, train_loader, test_loader):
 
  # testing phase
  model.eval()
  predict_train = []
  predict_test = []
  labels_train = []
  labels_test = []

  with torch.no_grad():
      for inputs, labels in train_loader:
          outputs = model(inputs)
          index_, predicted = torch.max(outputs, dim=1)
          predict_train.append(predicted.tolist())
          labels_train.append(labels.tolist())

      for inputs, labels in test_loader:
          outputs = model(inputs)
          index_, predicted = torch.max(outputs, dim=1)
          predict_test.append(predicted.tolist())
          labels_test.append(labels.tolist())

  print("Confusion matrix on train:\n",  confusion_matrix(list(chain(*labels_train)), list(chain(*predict_train)), labels=[0, 1]))
  print()
  print("Classification report on train:\n",  classification_report(list(chain(*labels_train)), list(chain(*predict_train)), labels=[0, 1]))
  print()
  print("Confusion matrix on test:\n",  confusion_matrix(list(chain(*labels_test)), list(chain(*predict_test)), labels=[0, 1]))
  print()
  print("Classification report on test:\n",  classification_report(list(chain(*labels_test)), list(chain(*predict_test)), labels=[0, 1]))


In [156]:
# Examine evaluation results
test(model,train_loader,test_loader)

Confusion matrix on train:
 [[2572    0]
 [   0  515]]

Classification report on train:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      2572
           1       1.00      1.00      1.00       515

    accuracy                           1.00      3087
   macro avg       1.00      1.00      1.00      3087
weighted avg       1.00      1.00      1.00      3087


Confusion matrix on test:
 [[1122    5]
 [  10  186]]

Classification report on test:
               precision    recall  f1-score   support

           0       0.99      1.00      0.99      1127
           1       0.97      0.95      0.96       196

    accuracy                           0.99      1323
   macro avg       0.98      0.97      0.98      1323
weighted avg       0.99      0.99      0.99      1323



The model predicts 191 cases of attrition, out of which 186 employees actually have left the company. There are 5 false negative, what gives us the best precision score = 0.97. Model also didn't realize just 10 of employees who actually left the company, hence very good recall score = 0.95

In [147]:
!jupyter nbconvert --to html "/content/drive/MyDrive/DL_lab/ProjectV3.ipynb"

[NbConvertApp] Converting notebook /content/drive/MyDrive/DL_lab/ProjectV3.ipynb to html
[NbConvertApp] Writing 351748 bytes to /content/drive/MyDrive/DL_lab/ProjectV3.html
